Services/ExternalDataProviders/GithubDataSourceAdaptee.cs

```diff
@@ -0,0 +1,378 @@
+        /// </summary>
+        /// <param name="identifier">The identifier which is used to retrieve the correct project.</param>
+        /// <returns>This method returns a github data source resource result.</returns>
+        /// <exception cref="ExternalException">This method could throw an external exception whenever the status code is not successful.</exception>
+        private async Task<GithubDataSourceResourceResult> FetchPublicGithubProjectById(string identifier)
+        {
+            IRestClient client = restClientFactory.Create(new Uri(BaseUrl));
+            IRestRequest request = new RestRequest($"repositories/{identifier}", Method.GET);
+            IRestResponse response = await client.ExecuteAsync(request);
+
+            if(string.IsNullOrEmpty(response.Content)) return null;
+            if(!response.IsSuccessful) throw new ExternalException(response.ErrorMessage);
+            GithubDataSourceResourceResult resourceResult =
+                JsonConvert.DeserializeObject<GithubDataSourceResourceResult>(response.Content);
+            return resourceResult;
+        }
+
+        /// <summary>
+        /// This method is responsible for retrieving the readme from a repository.
+        /// </summary>
+        /// <param name="user">This parameter represents the owners of the repository. This is in most cases the name of the user
+        /// or in some other case the name of the organization.</param>
+        /// <param name="repository">This parameter represents the name of the repository.</param>
+        /// <param name="accessToken">This parameter is the access token, which is used to retrieve a readme from a private
+        /// repository. In case of a public repository, the default value is null.</param>
+        /// <returns>This method returns the content of the readme.</returns>
+        /// <exception cref="ExternalException">This method could throw an external exception whenever the status code is not successful.</exception>
+        private async Task<string> FetchReadme(string user, string repository, string accessToken = null)
+        {
+            IRestClient client = restClientFactory.Create(new Uri(BaseUrl));
+            IRestRequest request = new RestRequest($"repos/{user}/{repository}/readme");
+
+            if(accessToken != null)
+            {
+                request.AddHeaders(new List<KeyValuePair<string, string>>
+                                   {
+                                       new KeyValuePair<string, string>("Authorization", $"Bearer {accessToken}"),
+                                       new KeyValuePair<string, string>("accept", "application/vnd.github.v3+json")
+                                   });
+            }
+
+            IRestResponse response = await client.ExecuteAsync(request);
+
+            if(!response.IsSuccessful) throw new ExternalException(response.ErrorMessage);
```

Comment:
If readme is not found GitHub returns a unsuccesful response. Consider to return null or "Readme not found" if Github throws NotFound.